changelogs/unreleased/gh-6253-net-box-iproto-id.md

```diff
@@ -0,0 +1,6 @@
+## feature/core
+
```

Comment:
As for the feature about requesting a certain version and features. Now looking at the patch, I am not sure it is needed((. Netbox could just support everything what it can depending on the tarantool version and `IPROTO_ID`, and raise errors on attempts to use the not defined features. For example, if you try to use `netbox:begin()` and the protocol does not support transactions, you will get an error anyway. Either netbox could raise it right away, or the server will. What do you think?

Maybe add it if somebody asks? Otherwise it complicates the netbox API a bit.

Reply:
There's a number of reasons why explicitly requiring and reporting peer features on connect can be useful:
1. Testing. Seeing peer features in net.box is the only convenient way to test `IPROTO_ID` until more complicated features are introduced.
2. Early bail out. True, `netbox:begin()` will raise an error if transactions are not supported by the server, but a client may do some time-consuming work before they issue the failing command. It'd be more efficient to bail out as soon as we detect that a feature required for application to function properly is not supported. Of course, a client could invoke all commands they are going to use before starting to actually doing anything, but this wouldn't be really user-friendly.
3. Probing commands may increment the error counter or spam the log on the server. Requiring features on connect is more graceful.
4. There may be features that may affect the application behavior without introducing new commands to the protocol. For example, consider error marshaling, which is added to the feature set by https://github.com/tarantool/tarantool/pull/6435. Instead of handling both string errors and error objects in their application, a client may require the server to support error marshaling and refuse to work if it doesn't. This way, the client can write their application assuming that errors returned by functions called on the server are complex objects.